pubtools/_pulp/tasks/publish.py

```diff
@@ -0,0 +1,190 @@
+import logging
+import sys
+import re
+from datetime import datetime
+from functools import partial
+from more_executors.futures import f_map, f_sequence
+
+from pubtools.pulplib import Criteria, Matcher, PublishOptions
+
+from pubtools._pulp.task import PulpTask, CDNCached
+from pubtools._pulp.services import PulpClientService, UdCacheClientService
+
+
+step = PulpTask.step
+
+LOG = logging.getLogger("pubtools.pulp")
+
+
+def publish_date(str_date):
+    # validates publish-before date
+    return datetime.strptime(str_date, "%Y-%m-%d")
+
+
+class Publish(PulpClientService, UdCacheClientService, PulpTask, CDNCached):
+    """Publishes the pulp repositories to the endpoints defined by the distributors
+
+    This command will publish the pulp repositories provided in the request or
+    fetched using the filters(url-regex or published-before) or an intersection
+    of input repos and filters.
+    """
+
+    def add_args(self):
+        super(Publish, self).add_args()
+
+        self.parser.add_argument(
+            "--repo-ids", help="list of repos to be published", nargs="+"
+        )
+        self.parser.add_argument(
+            "--clean",
+            help="attempt to remove content not in the repo",
+            action="store_true",
+        )
+        self.parser.add_argument(
+            "--force", help="execute a full repo publish", action="store_true"
+        )
+        self.parser.add_argument(
+            "--published-before",
+            help="publish the repos last published before given date e.g. 2019-08-21",
+            type=publish_date,
+            default=None,
+        )
+        self.parser.add_argument(
+            "--repo-url-regex",
+            help="publish repos whose repo url match",
+            default=None,
+            type=re.compile,
+        )
+
+    def run(self):
+        to_await = []
+        LOG.info("run publish")
+
+        # get repos applying filters
+        repos = self.get_repos()
+
+        # publish the repos found
+        publish_fs = self.publish(repos)
+
+        # wait for the publish to complete before
+        # flushing caches.
+        f_sequence(publish_fs).result()
+
+        # flush UD cache
+        to_await.extend(self.flush_ud(repos))
+
+        # flush CDN cache
+        to_await.extend(self.flush_cdn(repos))
+
+        # wait for everthing to finish.
+        for f in to_await:
+            f.result()
+
+        LOG.info("Publishing repositories completed")
+
+    @step("Publish")
+    def publish(self, repos):
+        out = []
+
+        publish_opts = PublishOptions(force=self.args.force, clean=self.args.clean)
+        for repo in repos:
+            LOG.info("Publishing %s", repo.id)
+            f = repo.publish(publish_opts)
+            f = f_map(f, partial(self.log_publish, repo))
+            out.append(f)
+
+        return out
+
+    def log_publish(self, repo, tasks):
+        # logs publish status
+        for task in tasks:
```

Comment:
I don't think the logic here quite makes sense:

Most repos are going to have multiple distributors and so multiple tasks.  That means they're going to create multiple identical log messages in sequence, like:

      Publishing some-repo successful
      Publishing some-repo successful

Also I don't think you need to check task.succeeded because the API is already defined as only returning the tasks if they've succeeded (otherwise the future will raise an exception).

I'm thinking actually you don't need to log anything here at all and you can let the pubtools-pulplib task completion log messages cover it.  We discussed something similar in a maintenance-mode code reviews - currently those log messages may not contain all needed details, but it seems we can easily add them (task tags).

Reply:
Here's a PR extending the log messages in pulplib: https://github.com/release-engineering/pubtools-pulplib/pull/69

It would give you messages like:

     Pulp task completed: a1b2c3d4, repository:some-repo, action:publish

So I think if that PR is merged, then this logging could be dropped entirely.